platforms/common/include/px4_platform_common/atomic_bitset.h

```diff
@@ -0,0 +1,98 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2020 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#pragma once
+
+#include "atomic.h"
+
+namespace px4
+{
+
+template <size_t N>
+class AtomicBitset
+{
+public:
+
+	AtomicBitset() = default;
+
+	size_t count() const
+	{
+		size_t total = 0;
+
+		for (const auto &x : _data) {
+			const uint32_t y = x.load();
+
+			for (uint8_t i = 0; i < BITS_PER_ELEMENT; i++) {
```

Comment:
This is much faster:
```suggestion
v = v - ((v >> 1) & 0x55555555);                    // reuse input as temporary
v = (v & 0x33333333) + ((v >> 2) & 0x33333333);     // temp
c = ((v + (v >> 4) & 0xF0F0F0F) * 0x1010101) >> 24; // count
```
http://graphics.stanford.edu/~seander/bithacks.html#CountBitsSetParallel

Reply:
I'm aware of this one (I believe it's what llvm & gcc use for the popcount builtin), but I didn't want to go there unless we really needed it.

Reply:
Why not? The code is already based on 32 bit integers.

Reply:
It's a few lines of borderline magic to optimize something that I don't think is even being called outside of the unit tests (so far). If/when we need it I'd still want to consider using `__builtin_popcount` first.

Reply:
Well, this still hurts my eyes a bit. Can you remove the condition in the loop?
```
while(y) {
  total += y & 1;
  y >>= 1;
}
```

Reply:
Sure.